### PR TITLE
1076 autogenerated names

### PR DESCRIPTION
**Pull request by hwindsor**
## Pull Request Overview

- Implementation of Autogenerated names controlled from Site settings
- Individuals can automatically have autogenerated names for the species.
- A bit more site stting refactoring to make it more scalable for odd cases such as this one.

---
See doc : https://docs.google.com/document/d/1tHhuixb_TY5cYhwuy4qZKrZafvFyti8qac_RRSki9NA/edit#

**Model Updates **

`AutogeneratedName`
- `guid`
- `type`
- `prefix`
- `reference guid` 

**REST API Updates **

Site settings main block get andpost now also contain the following field 
```
{
    'GUID': {
        'type':  'auto-species',   # mandatory
        'prefix': 'indo',          # mandatory
        'reference_guid': 'GUID',  # optional
    }
}
```

### CODE REVIEW

File: app/modules/autogenerated_names/models.py [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+            'reference_guid={self.reference_guid}'
+            ')>'.format(class_name=self.__class__.__name__, self=self)
+        )
+
+    def get_next(self):
+        next_val = str(self.next_value).zfill(4)
+
+        with db.session.begin():
+            # Update and write as close as possible to minimise race condition seen in old Wildbook
+            self.next_value += 1
```
Comment (hwindsor) [inline app/modules/autogenerated_names/models.py:58]: tight enough for you @naknomum ?

Comment (naknomum) [reply app/modules/autogenerated_names/models.py:63]: i think it works -- for now.  :grin:   more importantly, you have a nice little method we can change if it proves not to be.  :heavy_check_mark:

File: app/modules/individuals/models.py [hunk 2]
```diff
@@ -383,6 +383,17 @@ def remove_names_for_value(self, value):
...
+    def add_autogenerated_species_name(self, user):
+        from app.modules.autogenerated_names.models import AutogeneratedName
+
+        tx_guid = self.get_taxonomy_guid()
+        if not tx_guid:
+            return
+        autogen_names = AutogeneratedName.query.filter_by(reference_guid=tx_guid).all()
+        if autogen_names and len(autogen_names) == 1:
+            new_name = autogen_names[0].get_next()
+            self.add_name('autogen-species', new_name, user)
```
Comment (naknomum) [inline app/modules/individuals/models.py:395]: not a blocking comment, but i wonder if this string used for the context on the actual name added should be derived from the `AutogeneratedNameType` values, like `auto_species`; just so we have a way to connect between the two without needing a new set of values for the name-context.  (tho i guess this value could be algorithmically generated from the AutogeneratedNameType, like `autogen-` + text after the `_`.)

**Comment by bluemellophone** [comment]
Yes, this PR is meant to be the first stab so the FE can start iterating.  The intention is to have a follow-up PR to update this feature based on where FE takes it.

**Pull request merged**